### PR TITLE
DEVEXP-307: Conversations Management APIs 

### PR DESCRIPTION
**Pull request by JPPortier**
Note: InjectMessage/InjectEvent not part of this PR

### CODE REVIEW

File: client/src/main/com/sinch/sdk/domains/conversation/models/v1/conversation/request/InjectMessageRequestImpl.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  protected InjectMessageRequestImpl(
+      OptionalValue<AppMessage<?>> appMessage,
+      OptionalValue<ContactMessage<?>> contactMessage,
+      OptionalValue<Instant> acceptTime,
+      OptionalValue<ChannelIdentity> channelIdentity,
+      OptionalValue<String> contactId,
+      OptionalValue<ConversationDirection> direction,
+      OptionalValue<String> conversationId,
+      OptionalValue<Boolean> injected,
```
Comment (asein-sinch) [inline client/src/main/com/sinch/sdk/domains/conversation/models/v1/conversation/request/InjectMessageRequestImpl.java:32]: Can you double check this parameter is part of an injectMessage **request**?

File: client/src/test/java/com/sinch/sdk/domains/conversation/api/v1/adapters/ConversationsServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    ConversationsListResponse response1 = response0.nextPage();
+    content = response1.getContent();
+    Assertions.assertThat(content.size()).isEqualTo(1);
+    TestHelpers.recursiveEquals(
+        content.stream().findFirst().orElse(null), ConversationDtoTest.conversation);
+    Assertions.assertThat(response1.hasNextPage()).isEqualTo(true);
+
+    ConversationsListResponse response2 = response1.nextPage();
+    content = response2.getContent();
+    Assertions.assertThat(content.size()).isEqualTo(0);
```
Comment (asein-sinch) [inline client/src/test/java/com/sinch/sdk/domains/conversation/api/v1/adapters/ConversationsServiceTest.java:167]: Nice thing to test this edge-case for the pagination of this API!

File: client/src/test/java/com/sinch/sdk/domains/conversation/api/v1/adapters/ConversationsServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            eq(ConversationDtoTest.conversation),
+            eq(null),
+            eq(MetadataUpdateStrategy.MERGE_PATCH)))
+        .thenReturn(ConversationDtoTest.conversation);
+
+    Conversation response =
+        service.update(
+            ConversationDtoTest.conversation.getId(),
+            MetadataUpdateStrategy.MERGE_PATCH,
+            ConversationDtoTest.conversation);
```
Comment (asein-sinch) [inline client/src/test/java/com/sinch/sdk/domains/conversation/api/v1/adapters/ConversationsServiceTest.java:248]: Indeed, when updating a conversation, you can add the `id` but it won't be taken into account

File: openapi-contracts/src/test/java/com/sinch/sdk/domains/conversation/models/v1/conversations/request/InjectMessageDtoTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+          .setDirection(ConversationDirection.TO_CONTACT)
+          .setConversationId("conversation id")
+          .setInjected(true)
+          .setSenderId("send id value")
+          .setProcessingMode(ProcessingMode.CONVERSATION)
+          .setMetadata("metdata value")
+          .build();
+  public static InjectMessageRequest injectContactMessage =
+      InjectMessageRequest.<CardMessage>builder()
+          .setBody(ContactMessageDtoTest.textContactMessageDto)
```
Comment (asein-sinch) [inline openapi-contracts/src/test/java/com/sinch/sdk/domains/conversation/models/v1/conversations/request/InjectMessageDtoTest.java:36]: Nit: naming `appTextMessageDto` vs `textContactMessageDto` 🤓

File: openapi-contracts/src/test/resources/domains/conversation/v1/conversations/response/ConversationsListResponse-page-1.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      "last_received": "2020-11-17T15:00:00Z",
+      "metadata": "metadata value",
+      "metadata_json": {
+        "metadata_json_key": "metadata json value"
+      },
+      "correlation_id": "correlation id value"
+    }
+  ],
+  "next_page_token": "ChowMUo1NUVBSk1FNDZQMlI4SFIxRjVZRjYxWhIaMDFITkIzMzY0UFdCQkcyR1paV1ZWVDNERDI=",
+  "total_size": 0
```
Comment (asein-sinch) [inline openapi-contracts/src/test/resources/domains/conversation/v1/conversations/response/ConversationsListResponse-page-1.json:18]: What a weird way to handle pagination! 😣

File: client/src/main/com/sinch/sdk/domains/conversation/models/v1/conversation/request/InjectMessageRequestImpl.java (hunk 1 above)
Comment (JPPortier) [reply client/src/main/com/sinch/sdk/domains/conversation/models/v1/conversation/request/InjectMessageRequestImpl.java:32]: Good catch/question.
Currently specs claim it is part of schema for request yes.
May be a refinement of specs could be imagine but having to check with product team.
Currently:
- at worst the field will be ignored by server
- at best: specs are right and user can set this field itself.

Question asked to product's team. A new PR could be created to synch if specs will be updated

Comment (asein-sinch) [reply client/src/main/com/sinch/sdk/domains/conversation/models/v1/conversation/request/InjectMessageRequestImpl.java:32]: Thanks for your reactivity. On the documentation side, the field is not present 🤨
https://developers.sinch.com/docs/conversation/api-reference/conversation/tag/Conversation/#tag/Conversation/operation/Conversation_InjectMessage

**Review by asein-sinch — APPROVED** [review]

File: client/src/test/java/com/sinch/sdk/domains/conversation/api/v1/adapters/ConversationsServiceTest.java (hunk 3 above)
Comment (JPPortier) [reply client/src/test/java/com/sinch/sdk/domains/conversation/api/v1/adapters/ConversationsServiceTest.java:248]: Yes and unit test are based onto specs definitions: specs says a Conversation instance can be send to server and then with an id even if ignored... 
If specs are refined, this mean unit tests will have to be updated and this is the purpose of such tests 😎

File: openapi-contracts/src/test/java/com/sinch/sdk/domains/conversation/models/v1/conversations/request/InjectMessageDtoTest.java (hunk 4 above)
Comment (JPPortier) [reply openapi-contracts/src/test/java/com/sinch/sdk/domains/conversation/models/v1/conversations/request/InjectMessageDtoTest.java:36]: Good catch: removed unnecessary code: InjectMessageRequest is not based onto generics

Comment (JPPortier) [reply openapi-contracts/src/test/java/com/sinch/sdk/domains/conversation/models/v1/conversations/request/InjectMessageDtoTest.java:36]: Fixed
